Mathlib/AlgebraicGeometry/EllipticCurve/Projective.lean

```diff
@@ -446,4 +450,716 @@ alias nonsingular_of_affine_of_Z_ne_zero := nonsingular_of_Z_ne_zero
 @[deprecated (since := "2024-08-27")] alias nonsingular_smul_iff := nonsingular_smul
 @[deprecated (since := "2024-08-27")] alias nonsingular_zero' := nonsingular_zero
 
+section Negation
+
+/-! ### Negation formulae -/
+
+variable (W') in
+/-- The $Y$-coordinate of a representative of `-P` for a point `P`. -/
+def negY (P : Fin 3 → R) : R :=
+  -P y - W'.a₁ * P x - W'.a₃ * P z
+
+lemma negY_eq (X Y Z : R) : W'.negY ![X, Y, Z] = -Y - W'.a₁ * X - W'.a₃ * Z :=
+  rfl
+
+lemma negY_smul (P : Fin 3 → R) (u : R) : W'.negY (u • P) = u * W'.negY P := by
+  simp only [negY, smul_fin3_ext]
+  ring1
+
+lemma negY_of_Z_eq_zero [NoZeroDivisors R] {P : Fin 3 → R} (hP : W'.Equation P) (hPz : P z = 0) :
```

Comment:
You can write `variable {P Q : Fin 3 → F}`

Reply:
Some of these only require the ring to be commutative (possibly with `NoZeroDivisors`), while others need the ring to be a field, so I thought the variables `P` and `Q` should be instantiated along with the lemma.